proposals/0000-quick-look-impredicativity.md

```diff
@@ -0,0 +1,181 @@
+
+`(Int -> Int) -> R is a subtype of (forall a. a -> a) -> R`
+
+The problem is of technical nature, but without this restriction we would not be able to infer impredicative instantiations of functions like `($)`. **This restriction affects not only users of `ImpredicativeTypes`, but also users of `RankNTypes`, effectively making this proposal not 100% backwards compatible.**
+
+Note anyway that the support for contravariance in GHC is really fragile. Take for example the functions `choose :: forall a. (a -> Bool) -> (a -> Bool) -> Bool`, `f :: (Bool -> Bool) -> Bool` and `g :: (forall a. a -> a) -> Bool`. In theory we can accept the expressions `choose f g` and `choose g f` if we instantiate `a` with `(forall a. a -> a)`. Here is an excerpt of a GHCi session which shows that the one of the expressions is accepted and the other is not:
+
+```text
+Prelude> :set -XRankNTypes -XImpredicativeTypes
+Prelude> choose :: (a -> Bool) -> (a -> Bool) -> Bool ; choose = undefined
+Prelude> f :: (Bool -> Bool) -> Bool ; f = undefined
+Prelude> g :: (forall a. a -> a) -> Bool ; g = undefined
+Prelude> :t choose f g
+
+<interactive>:1:10: error:
+    • Couldn't match type ‘Bool -> Bool’ with ‘forall a. a -> a’
+      Expected type: (Bool -> Bool) -> Bool
+        Actual type: (forall a. a -> a) -> Bool
+    • In the second argument of ‘choose’, namely ‘g’
+      In the expression: choose f g
+Prelude> :t choose g f
+choose g f :: Bool
+```
+
+There is a simple workaround that works most of the time: eta-expansion (you might need to eta-expand more than once, though). Taking the `choose`, `f`, and `g` functions defined above,in a non-contravariant world we need to eta-expand `f` to make it work: `choose (\x -> f x) g`; now the `x` is assigned type `forall a. a -> a`, which is then instantiated to `Int -> Int` before passing it to `f`.
+
+### The `ContravariantFunctions` extension
+
+After a preliminary assessment of which packages still compile under this new policy, it looks like most of them do not need it. Except one big and important one: `Cabal` (and anything that uses its API). This library often uses functions of the form:
+
+```haskell
+f :: ( (... -> CabalIO a) -> ...) -> ...
+```
+
+where `CabalIO a` is defined as `HasCallStack => IO a`. That means that whereas before we could pass a function `g` as argument, now we have to do something akin to `\... t ... -> g (\... x ... -> t ... x ...)`, which is quite annoying.
+
+We thus propose to add another extension to GHC, called `ContravariantFunctions`, which should be disabled by default. When the extension is enabled, subsumption checking is done using full contravariance of function types (but during quick look contravariance does not enter the game). This means that `RankNTypes` + `ContravariantFunctions` behaves the same as the old `RankNTypes. This has the benefit of asking people to eta-expand their functions, while having a escape way for those libraries which would require too many changes.
```

Comment:
If that's part of the proposal, `-XContravariantFunctions` should be in the _Specification_ section. If it is not, it should be in the _Alternatives_ section.

> but during quick look contravariance does not enter the game

I don't understand this sentence. Does it mean that `-XContravariantFunctions` is incompatible with `-XImpredicativeTypes`? Or something else?

Reply:
Moved to the "Alternatives" section. The sentence you quoted does not make much sense at that point, so I have removed it.